File: src/Day03.hs
module Day03 where

import AOCSolution (getSolution)
import Common
import Data.Char (digitToInt)
import Data.Function (on)
import Data.List (maximumBy)
import Inputs (InputType (..), readInput)

day03SampleInput, day03ActualInput :: IO (Maybe String)
day03SampleInput = readInput Sample 3
day03ActualInput = readInput Actual 3

day03 :: IO ((String, String), (String, String))
day03 = do
  s <- maybe ("", "") solve <$> day03SampleInput
  a <- maybe ("", "") solve <$> day03ActualInput
  return (s, a)

-- >>> day03
-- (("357","3121910778619"),("17085","169408143086082"))

solve :: String -> (String, String)
solve = getSolution p (f 2) (f 12)
  where
    f n = sum . map (makeBest n)
    p = map (map digitToInt) . lines

makeBest :: Int -> [Int] -> Int
makeBest n vals = intListToInt $ take n $ go initN vals
  where
    initN = length vals - n + 1
    go :: Int -> [Int] -> [Int]
    go _ [] = []
    go n xs = v : go n' xs''
      where
        (v, xs') = splitAtMax (take n xs)
        n' = length xs' + 1
        xs'' = xs' ++ drop n xs

splitAtMax :: [Int] -> (Int, [Int])
splitAtMax xs = go (0, xs) xs
  where
    go (curMax, after) [] = (curMax, after)
    go (curMax, after) (y : ys)
      | y > curMax = go (y, ys) ys
      | otherwise = go (curMax, after) ys

File: src/day03.rs
use std::fs;

/// Sample
/// ```
/// use advent_of_code_2025::{check_day_output, day03::run};
/// check_day_output!(run, "sample", "357", "3121910778619");
/// ```
///
/// Actual
/// ```
/// use advent_of_code_2025::{check_day_output, day03::run};
/// check_day_output!(run, "actual", "17085", "169408143086082");
/// ```
pub fn run(file_type: &str) -> Option<(String, String)> {
    let contents = fs::read_to_string(format!("inputs/{}/03.txt", file_type))
        .expect("Should have been able to read the file");

    let input = &parse_input(&contents);

    Some((solve(input, 2).to_string(), solve(input, 12).to_string()))
}

fn parse_input(input: &str) -> Vec<Vec<u64>> {
    input
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| c.to_digit(10).expect("Should be able to parse digit") as u64)
                .collect::<Vec<u64>>()
        })
        .collect::<Vec<Vec<u64>>>()
}

fn solve(input: &Vec<Vec<u64>>, length: usize) -> u64 {
    input
        .iter()
        .map(|bank| make_best(length, bank))
        .sum()
}

fn make_best(length: usize, vals: &Vec<u64>) -> u64 {
    let mut to_remove = vals.len() - length;
    let mut best_digits: Vec<u64> = Vec::with_capacity(vals.len());

    for &digit in vals {
        while to_remove > 0
            && best_digits
                .last()
                .is_some_and(|&last_digit| last_digit < digit)
        {
            best_digits.pop();
            to_remove -= 1;
        }

        best_digits.push(digit);
    }

    best_digits.truncate(length);

    best_digits
        .into_iter()
        .fold(0_u64, |acc, digit| acc * 10 + digit)
}
